lte/gateway/c/li_agent/PDUGenerator.cpp

```
@@ -0,0 +1,166 @@
+    const struct pcap_pkthdr* phdr, struct conditional_attributes* attributes) {
+  attributes->timestamp = phdr->ts.tv_sec;
+}
+
+void* PDUGenerator::generate_pkt(
+    const struct pcap_pkthdr* phdr, const u_char* pdata) {
+  uint8_t* data = (uint8_t*) calloc(
+      1, sizeof(struct pdu_info) + sizeof(struct conditional_attributes) +
+             phdr->len);
+  struct pdu_info* pdu = (struct pdu_info*) data;
+
+  pdu->version  = PDU_VERSION;
+  pdu->pdu_type = PDU_TYPE;
+  pdu->header_length =
+      sizeof(struct pdu_info) + sizeof(struct conditional_attributes);
+  pdu->payload_length = phdr->len;
+  pdu->payload_format = IP_PAYLOAD_FORMAT;
+
+  set_conditional_attr(
+      phdr, (struct conditional_attributes*) (data + sizeof(struct pdu_info)));
+
+  memcpy(data + pdu->header_length, pdata, phdr->len);
+
+  MLOG(MDEBUG) << "Generated packet length with length - "
+               << pdu->header_length + pdu->payload_length;
+  print_bytes(data, pdu->header_length + pdu->payload_length);
+
+  return (void*) data;
+}
+
+bool PDUGenerator::send_packet(
+    const struct pcap_pkthdr* phdr, const u_char* pdata) {
+  std::string src_ip;
+  std::string dst_ip;
+  if (!extract_ip_addr(pdata, src_ip, dst_ip)) {
+    MLOG(MERROR) << "Could not extract IP from the packet, skipping";
+    return true;
+  }
+
+  MLOG(MDEBUG) << "Processing packet with src ip " << src_ip << " dst ip "
+               << dst_ip;
+  void* data           = generate_pkt(phdr, pdata);
+  struct pdu_info* pdu = (struct pdu_info*) data;
+
+  PacketSender sender;
+  directoryd_client_->get_directoryd_xid_field(
+      src_ip, [this, src_ip, data, pdu](Status status, DirectoryField resp) {
+        // TODO process the output of the directoryd lookup
+        pdu->payload_direction = DIRECTION_FROM_TARGET;
+
+        set_xid(pdu, "tracking_123");
+        proxy_connector_->SendData(
+            data, pdu->header_length + pdu->payload_length);
+
+        if (!status.ok()) {
```

Related to my comment elsewhere on Result and Status type needs in Magma C++... Are you swallowing an error here and should this function return false?